File: windows/src/SkillHub.Core/HubPaths.cs
namespace SkillHub.Core;

public sealed class HubPaths : IEquatable<HubPaths>
{
    public static readonly HashSet<string> IgnoreNames =
    [
        "node_modules", ".git", ".raw", "vendor_imports", ".tmp",
        "plugins.bak", "dist", ".DS_Store", "DerivedData", "build",
        "test-results", "comparison-results"
    ];

    public string Home { get; }

    public HubPaths(string home)
    {
        Home = PathUtil.Standardize(home);
    }

    public string ContentRoot => Path.Combine(Home, ".skill-hub");
    public string ArchiveRoot => Path.Combine(ContentRoot, "archive");
    public string PromptLibrary => Path.Combine(ContentRoot, "library", "prompts");
    public string LegacyMetaFile => Path.Combine(ContentRoot, "meta.json");

    public string? SkillRoot(ToolSource source)
    {
        var relative = source switch
        {
            ToolSource.CursorUser => Path.Combine(".cursor", "skills"),
            ToolSource.CursorBuiltin => Path.Combine(".cursor", "skills-cursor"),
            ToolSource.Claude => Path.Combine(".claude", "skills"),
            ToolSource.Codex => Path.Combine(".codex", "skills"),
            ToolSource.Agents => Path.Combine(".agents", "skills"),
            ToolSource.Proma => Path.Combine(".proma", "default-skills"),
            _ => null
        };
        return relative == null ? null : Path.Combine(Home, relative);
    }

    public string? PromptRoot(ToolSource source) => source switch
    {
        ToolSource.PromptLibrary => PromptLibrary,
        ToolSource.CodexPrompts => Path.Combine(Home, ".codex", "prompts"),
        _ => null
    };

    public bool IsBuiltinPath(string path)
    {
        var builtin = SkillRoot(ToolSource.CursorBuiltin);
        if (builtin == null)
        {
            return false;
        }

        return PathUtil.IsUnder(builtin, path);
    }

    public void EnsureContentDirectories()
    {
        Directory.CreateDirectory(PromptLibrary);
        Directory.CreateDirectory(ArchiveRoot);
    }

    public IReadOnlyList<string> ProjectSkillRoots()
    {
        var seen = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        var seenProjectFolders = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        var roots = new List<string>();
        foreach (var folderName in new[] { "Projects", "projects" })
        {
            var projects = Path.Combine(Home, folderName);
            if (!Directory.Exists(projects))
            {
                continue;
            }

            if (!seenProjectFolders.Add(RealOrStandard(projects)))
            {
                continue;
            }

            foreach (var child in PathUtil.ImmediateChildren(projects))
            {
                if (!Directory.Exists(child))
                {
                    continue;
                }

                foreach (var relative in new[] { Path.Combine(".cursor", "skills"), Path.Combine(".claude", "skills") })
                {
                    var candidate = Path.Combine(child, relative);
                    if (!Directory.Exists(candidate))
                    {
                        continue;
                    }

                    var real = PathUtil.RealPath(candidate);
                    if (seen.Add(real))
                    {
                        roots.Add(candidate);
                    }
                }
            }
        }

        return roots;
    }

    public IReadOnlyList<string> WritableRoots(ScanOptions options)
    {
        var roots = ToolSourceInfo.SkillRoots
            .Select(SkillRoot)
            .Where(p => p != null)
            .Select(p => PathUtil.Standardize(p!))
            .ToList();
        roots.Add(PathUtil.Standardize(ContentRoot));
        roots.AddRange(options.PromptRoots.Select(PathUtil.Standardize));
        var codex = PromptRoot(ToolSource.CodexPrompts);
        if (codex != null)
        {
            roots.Add(PathUtil.Standardize(codex));
        }

        if (options.ScanProjectSkills)
        {
            roots.AddRange(ProjectSkillRoots().Select(PathUtil.Standardize));
        }

        return roots;
    }

    static string RealOrStandard(string path)
    {
        try
        {
            return PathUtil.RealPath(path);
        }
        catch (IOException)
        {
            return PathUtil.Standardize(path);
        }
    }

    public bool Equals(HubPaths? other) => other != null && Home == other.Home;
    public override bool Equals(object? obj) => obj is HubPaths other && Equals(other);
    public override int GetHashCode() => Home.GetHashCode();
}

File: SkillHub/Catalog/Services/HubPaths.swift
import Foundation

/// Every filesystem location Skill Hub reads or writes, derived from one home directory so the
/// same code works against the real home in the app and a temporary folder in tests.
nonisolated struct HubPaths: Sendable, Equatable {
    let home: URL

    init(home: URL) {
        self.home = home.standardizedFileURL
    }

    static let ignoreNames: Set<String> = [
        "node_modules", ".git", ".raw", "vendor_imports", ".tmp",
        "plugins.bak", "dist", ".DS_Store", "DerivedData", "build",
        "test-results", "comparison-results"
    ]

    /// User-visible content root. Prompts and archives live here so other tools can reach them.
    var contentRoot: URL { home.appending(path: ".skill-hub", directoryHint: .isDirectory) }
    var archiveRoot: URL { contentRoot.appending(path: "archive", directoryHint: .isDirectory) }
    var promptLibrary: URL { contentRoot.appending(path: "library/prompts", directoryHint: .isDirectory) }
    /// Location of the metadata file before it moved into the app container.
    var legacyMetaFile: URL { contentRoot.appending(path: "meta.json") }

    func skillRoot(_ source: ToolSource) -> URL? {
        let relative: String? = switch source {
        case .cursorUser: ".cursor/skills"
        case .cursorBuiltin: ".cursor/skills-cursor"
        case .claude: ".claude/skills"
        case .codex: ".codex/skills"
        case .agents: ".agents/skills"
        case .proma: ".proma/default-skills"
        case .promptLibrary, .codexPrompts, .custom: nil
        }
        return relative.map { home.appending(path: $0, directoryHint: .isDirectory) }
    }

    func promptRoot(_ source: ToolSource) -> URL? {
        switch source {
        case .promptLibrary: promptLibrary
        case .codexPrompts: home.appending(path: ".codex/prompts", directoryHint: .isDirectory)
        default: nil
        }
    }

    func isBuiltinPath(_ path: String) -> Bool {
        guard let builtin = skillRoot(.cursorBuiltin) else { return false }
        let prefix = builtin.standardizedFileURL.path
        return path == prefix || path.hasPrefix(prefix + "/")
    }

    func ensureContentDirectories() throws {
        let fm = FileManager.default
        try fm.createDirectory(at: promptLibrary, withIntermediateDirectories: true)
        try fm.createDirectory(at: archiveRoot, withIntermediateDirectories: true)
    }

    /// `.cursor/skills` and `.claude/skills` folders inside each project under `~/Projects`.
    func projectSkillRoots() -> [URL] {
        let fm = FileManager.default
        var seen = Set<String>()
        var roots: [URL] = []
        for folderName in ["Projects", "projects"] {
            let projects = home.appending(path: folderName, directoryHint: .isDirectory)
            guard let children = try? fm.contentsOfDirectory(
                at: projects,
                includingPropertiesForKeys: [.isDirectoryKey],
                options: [.skipsHiddenFiles]
            ) else { continue }
            for child in children {
                for relative in [".cursor/skills", ".claude/skills"] {
                    let candidate = child.appending(path: relative, directoryHint: .isDirectory)
                    let real = Self.realPath(candidate)
                    guard fm.fileExists(atPath: candidate.path), seen.insert(real).inserted else { continue }
                    roots.append(candidate)
                }
            }
        }
        return roots
    }

    /// Roots the app is allowed to modify. Anything outside is refused before touching disk.
    func writableRoots(options: ScanOptions) -> [String] {
        var roots = ToolSource.skillRoots.compactMap { skillRoot($0)?.standardizedFileURL.path }
        roots.append(contentRoot.standardizedFileURL.path)
        roots.append(contentsOf: options.customPromptRoots.map { URL(fileURLWithPath: $0).standardizedFileURL.path })
        if let codex = promptRoot(.codexPrompts) {
            roots.append(codex.standardizedFileURL.path)
        }
        if options.scanProjectSkills {
            roots.append(contentsOf: projectSkillRoots().map { $0.standardizedFileURL.path })
        }
        return roots
    }

    static func standardizedPath(_ url: URL) -> String {
        url.standardizedFileURL.path
    }

    static func realPath(_ url: URL) -> String {
        url.resolvingSymlinksInPath().standardizedFileURL.path
    }
}
